x/cronos/keeper/precompiles/base_contract.go

```diff
@@ -10,18 +12,56 @@ type Registrable interface {
 
 type BaseContract interface {
 	Registrable
+	RequiredGas(input []byte) uint64
 }
 
 type baseContract struct {
-	address common.Address
+	address                       common.Address
+	kvGasConfig                   storetypes.GasConfig
+	nameByMethod                  map[[4]byte]string
+	gasByMethod                   map[[4]byte]uint64
+	emptyGasIfInputLessThanPrefix bool
+	logger                        log.Logger
 }
 
-func NewBaseContract(address common.Address) BaseContract {
+func NewBaseContract(
+	address common.Address,
+	kvGasConfig storetypes.GasConfig,
+	nameByMethod map[[4]byte]string,
+	gasByMethod map[[4]byte]uint64,
+	emptyGasIfInputLessThanPrefix bool,
+	logger log.Logger,
+) BaseContract {
 	return &baseContract{
-		address: address,
+		address,
+		kvGasConfig,
+		nameByMethod,
+		gasByMethod,
+		emptyGasIfInputLessThanPrefix,
+		logger,
 	}
 }
 
 func (c *baseContract) RegistryKey() common.Address {
 	return c.address
 }
+
+// RequiredGas calculates the contract gas use
+func (c *baseContract) RequiredGas(input []byte) (gas uint64) {
+	var methodID [4]byte
+	copy(methodID[:], input[:4])
+	inputLen := len(input)
+	defer func() {
+		method := c.nameByMethod[methodID]
+		c.logger.Info("required", "gas", gas, "method", method, "len", inputLen)
```

Comment:
what's this for, temporary debugging?

Reply:
I thought we want to record required gas for each method

Reply:
for debugging or what reason?

Reply:
yes debugging, we could collect the data on some nodes to tune the requiredgas map later on

but the data has to be the gas consumed at cosmos level during execution, not the value returned from the map

Reply:
I see, so this could be a draft PR for building a local node, no need to merge, right?

Reply:
I think so, could be just distributed to internal node

However this PR introduces a refactoring which could be nice to merge, maybe change the title of the PR , remove the log and just merge the refactoring?

Reply:
I tried compare gas used with cosmos and evm tx (current gas map vs estimate via relayer)

  | cosmos | evm_hardcode | evm_estimate
-- | -- | -- | --
MsgCreateClient | 121982 | 240794 | 123774
MsgUpdateClient+MsgConnectionOpenTry | 187503 | 841863 | 250059
MsgUpdateClient+MsgConnectionOpenConfirm | 145679 | 756978 | 177774
MsgUpdateClient+MsgChannelOpenTry | 205426 | 876532 | 240877
MsgUpdateClient+MsgChannelOpenConfirm | 154580 | 775420 | 188570
MsgUpdateClient+MsgRecvPacket | 280213 | 1163360 | 315581

Reply:
Thanks, do you know how cost a single MsgUpdateClient?

Based on that maybe we can use your estimation for better tuning

Reply:
since rly do auto update client, I get single MsgUpdateClient from hermes

  | herems
-- | --
MsgCreateClient | 122172
MsgConnectionOpenInit | 94511
MsgUpdateClient | 125791
MsgUpdateClient+MsgConnectionOpenAck | 178550
MsgChannelOpenInit | 143607
MsgUpdateClient+MsgChannelOpenAck | 156139
MsgUpdateClient+MsgRecvPacket | 279927

Reply:
I wonder if we could use those data on mainnet
https://cronos.yummy.capital/txs/E2B835B672BC046028E3FEBEB1770CD8A506DB505380982AD16C39F13E5DD861

Somehow hermes in mainnet is spending more gas than that. But may be also because of hermez config?

Reply:
we should have bigger client header like more validators, and I'm not sure why we use [TransientGasConfig](https://github.com/crypto-org-chain/cronos/blob/main/app/app.go#L548)

Reply:
For the TransientGasConfig, it inspired from evmos implementation of the `gasRequired` 
https://github.com/evmos/evmos/blob/main/precompiles/common/precompile.go#L34

They also use the `TransientGasConfig`, I think that make sense since the goal for us is just to add an "extra" fee based on the length of the input, to avoid DDOS attack for very large input

But in practice, the additional fee does not make much sense since input length is not related to the cost of operation in memory

Reply:
I think evmos use KvGasConfig instead and input length does related to [TxSizeCost](https://github.com/cosmos/cosmos-sdk/blob/v0.47.5/x/auth/ante/basic.go#L96) which is the major cost in cosmos tx

Reply:
mmm in the original context, the base cost is just an extra cost to avoid DDOS attack, since the input it never written in store, it would make sense to use transient store cost for this case.

But I think what you are trying to do is to replicate the cosmos intrinsic cost? In that case KvGasConfig is used